server/application-server/src/main/java/de/tum/cit/aet/helios/github/GitHubClientManager.java

```diff
@@ -0,0 +1,321 @@
+    PAT,
+    APP
+  }
+
+  public GitHubClientManager(String organizationName,
+                             String ghAuthToken,
+                             String appNameWithoutSuffix,
+                             Long appId,
+                             Long installationId,
+                             String privateKeyPath,
+                             OkHttpClient okHttpClient) {
+    this.organizationName = organizationName;
+    this.ghAuthToken = ghAuthToken;
+    this.appName = appNameWithoutSuffix + "[bot]";
+    this.appId = appId;
+    this.installationId = installationId;
+    this.privateKeyPath = privateKeyPath;
+    this.okHttpClient = okHttpClient;
+
+    if (appId != null && privateKeyPath != null) {
+      this.authType = AuthType.APP;
+    } else if (ghAuthToken == null || ghAuthToken.isEmpty()) {
+      this.authType = AuthType.PAT;
+    } else {
+      log.error(
+          "GitHub auth token or private key is not provided! "
+              + "GitHub client will be disabled.");
+      authType = null;
+    }
+  }
+
+  /**
+   * Returns the current GitHub token. It is either a PAT or an installation token.
+   *
+   * @return the current token
+   */
+  public String getCurrentToken() {
+    if (currentToken == null || Instant.now().isAfter(tokenExpirationTime)) {
+      refreshClient();
+    }
+    return currentToken;
+  }
+
+  /**
+   * Returns a GitHub client instance.
+   * If the client is null or the token has expired,
+   * a new client will be created. Generated client is valid for 1 hour.
+   * Every 20 minutes, the client will be refreshed.
+   *
+   * @return the GitHub client
+   */
+  public GitHub getGitHubClient() {
+    if (gitHubClient == null || Instant.now().isAfter(tokenExpirationTime)) {
+      refreshClient();
+    }
+    return gitHubClient;
+  }
+
+  /**
+   * Refreshes the GitHub client in every 20 minutes.
+   */
+  private void refreshClient() {
+    lock.lock();
+    try {
+      if (gitHubClient == null || Instant.now().isAfter(tokenExpirationTime)) {
+        log.info("Refreshing GitHub client...");
+        gitHubClient = createGitHubClientWithCache();
+        if (authType == AuthType.APP) {
+          fetchGitHubAppNodeId();
+        }
+        log.info("GitHub client refreshed successfully");
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Fetches the GitHub App node ID using the app name.
+   */
+  private void fetchGitHubAppNodeId() {
+    try {
+      githubAppNodeId = gitHubClient.getUser(appName).getNodeId();
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format("Failed to get GitHub Node ID of GitHub App with name: %s, error: %s",
+              appName, e.getMessage()), e);
+    }
+  }
+
+  /**
+   * Creates a GitHub client with caching enabled.
+   *
+   * @return the GitHub client
+   */
+  public GitHub createGitHubClientWithCache() {
+    if (authType == null) {
+      return GitHub.offline();
```

Comment:
suggestion: please add an warning log here, otherwise it is hard to debug why the github client is offline (already had this experience when fixing a bug)